src/Type/Constant/ConstantArrayType.php

```diff
@@ -602,11 +602,17 @@ public function findTypeAndMethodNames(): array
 
 	public function hasOffsetValueType(Type $offsetType): TrinaryLogic
 	{
-		$offsetType = $offsetType->toArrayKey();
+		$offsetArrayKeyType = $offsetType->toArrayKey();
+
+		return $this->recursiveHasOffsetValueType($offsetArrayKeyType);
+	}
+
+	private function recursiveHasOffsetValueType(Type $offsetType): TrinaryLogic
+	{
 		if ($offsetType instanceof UnionType) {
 			$results = [];
 			foreach ($offsetType->getTypes() as $innerType) {
-				$results[] = $this->hasOffsetValueType($innerType);
+				$results[] = $this->recursiveHasOffsetValueType($innerType);
```

Comment:
With the previous logic, toArrayKey was again called, creating a new UnionType for numeric string and an infinite loop.